### PR TITLE
8256073: Improve vector rematerialization support

### PR DESCRIPTION
**Pull request by iwanowww**
Having #1131, #1132, and #1134 in place, the only missing piece left to have vector rematerialization fully working is support of non-contiguous vector values in vector rematerialization logic. This patch covers that.

Current version makes the assumption that vector values are contiguously laid in memory. It's the case for on-stack locations, but for in-register values it's not the case (at least, on x86). Rewritten version doesn't make such assumption for in-register case anymore and processes every vector element independently.

(Along the way, the refactoring fixes a bug when handling a corner case: the case when a vector instance is scalarized by EA and the primitive array field (VectorPayload.payload) has a constant value (NULL) is erroneously treated as requiring custom rematerialization and it hits an assert.)

Testing (with other relevant patches):
- [x] jdk/incubator/vector w/ -XX:+DeoptimizeALot and -XX:UseAVX={3,2,1,0} on AVX512-capable hardware
- [x] hs-precheckin-comp, hs-tier1, hs-tier2

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux aarch64 | Linux arm | Linux ppc64le | Linux s390x | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- | ----- | ----- | ----- | ----- |
| Build | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (6/6 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) |    |     |     |     |  ❌ (1/9 failed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

**Failed test task**
- [Linux x64 (hs/tier1 runtime)](https://github.com/iwanowww/jdk/runs/1432305190)

### Issue
 * [JDK-8256073](https://bugs.openjdk.java.net/browse/JDK-8256073): Improve vector rematerialization support

### Reviewers
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to df58c86cd879dd190d26eaef2c525e83ac6a5588
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1136/head:pull/1136`
`$ git checkout pull/1136`

### CODE REVIEW

**Review by TobiHartmann — APPROVED** [review]
Looks reasonable to me.

File: src/hotspot/share/prims/vectorSupport.cpp [hunk 1]
```diff
@@ -83,126 +83,87 @@ jint VectorSupport::klass2length(InstanceKlass* ik) {
...
-        break;
-      }
-      default:
-        fatal("unsupported: %s", type2name(elem_bt));
+      case T_BYTE:   arr->bool_at_put(index,  (*(jbyte*)addr) != 0); break;
+      case T_SHORT:  arr->bool_at_put(index, (*(jshort*)addr) != 0); break;
+      case T_INT:    // fall-through
+      case T_FLOAT:  arr->bool_at_put(index,   (*(jint*)addr) != 0); break;
+      case T_LONG:   // fall-through
+      case T_DOUBLE: arr->bool_at_put(index,  (*(jlong*)addr) != 0); break;
```
Comment (vnkozlov) [inline src/hotspot/share/prims/vectorSupport.cpp:94]: Why you push `bool` value for everything?!!!

**Review by vnkozlov — CHANGES_REQUESTED** [review]

File: src/hotspot/share/prims/vectorSupport.cpp (hunk 1 above)
Comment (vnkozlov) [reply src/hotspot/share/prims/vectorSupport.cpp:94]: Okay. After reading code more I understand that array is boolean type for case when `is_mask` is `true`. So changes are correct.

Comment (vnkozlov) [reply src/hotspot/share/prims/vectorSupport.cpp:94]: May be you need separate method to avoid this confusion. The code is separate anyway.

**Review by vnkozlov — CHANGES_REQUESTED** [review]

**Comment by vnkozlov** [comment]
After thinking more on this a separate method could complicate changes. At least add a comment.

File: src/hotspot/share/prims/vectorSupport.cpp [hunk 2]
```diff
@@ -83,126 +83,87 @@ jint VectorSupport::klass2length(InstanceKlass* ik) {
...
+
+      default: assert(false, "unsupported: %s", type2name(elem_bt));
```
[the remark this answers is not in the document]
Comment (iwanowww) [reply src/hotspot/share/prims/vectorSupport.cpp]: Fixed.

**Comment by iwanowww** [comment]
> After thinking more on this a separate method could complicate changes. At least add a comment.

Added. Frankly speaking, I was on a fence when touching that code, but found it cleaner to have masks and vectors initialized in the same place.

File: src/hotspot/share/prims/vectorSupport.cpp [hunk 3]
```diff
@@ -83,126 +83,94 @@ jint VectorSupport::klass2length(InstanceKlass* ik) {
...
-  for (int i = 0; i < num_elem; i++) {
+void VectorSupport::init_payload_element(typeArrayOop arr, bool is_mask, BasicType elem_bt, int index, address addr) {
+  if (is_mask) {
+    // Masks require special handling: when boxed they are packed and stored in boolean
+    // arrays, but in scalarized form they have the same size as corresponding vectors.
+    // For example, Int512Mask is represented in memory as boolean[16], but
+    // occupies the whole 512-bit vector register when scalarized.
+    // (In generated code, the conversion is performed by VectorStoreMask.)
+    //
+    // TODO: revisit when predicate registers are fully supported.
```
Comment (vnkozlov) [inline src/hotspot/share/prims/vectorSupport.cpp:94]: Good and thanks.

**Review by vnkozlov — APPROVED** [review]

**Comment by iwanowww** [comment]
Thanks for the reviews, Tobias and Vladimir.

**Pull request closed without merging**